Add support for non_differentiable attribute

PhrygianGates opened the pull request:
This pull request introduces support for the 'non_differentiable' attribute applicable to fields, functions, and custom classes.

There are, however, a few lingering concerns I'll be addressing in the days to come. Some of the tests pass locally on my machine, but falter within the CI environment. This discrepancy might be related to differing versions of Clang or LLVM.

In addition, we might need to amend numerous 'visit' functions to facilitate the 'non_differentiable' attribute support for user-defined classes—this would include instances such as assignment statements and the like.

parth-07 commented on .gitignore:
```diff
@@ -1,2 +1,5 @@
 /build
 /.vscode
+/inst
+/workspace
```
Can you please add these files to your local .gitignore? If you think these git ignores are useful
for many other people as well, then can you please open a separate pull request to add these to .gitignore?

PhrygianGates replied:
okay I'll do that

parth-07 commented on lib/Differentiator/DerivativeBuilder.cpp:
```diff
@@ -179,6 +179,11 @@ namespace clad {
     FD = FD->getDefinition();
+    auto* Attr = FD->getAttr<clang::AnnotateAttr>();
+    // Skip functions marked with 'annotate("non_differentiable")'.
+    if (Attr && Attr->getAnnotation() == "non_differentiable") {
```
We should not just skip this. We should report an error, if a user has explicitly requested to differentiate a non-differentiable marked function.

PhrygianGates replied:
ok

parth-07 replied:
Can you please add a test case in which a user tries to request a derivative of a function (using `clad::differentiate`) that is explicitly marked as `non_differentiable`?

parth-07 commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -1550,6 +1625,14 @@ namespace clad {
     for (auto D : DS->decls()) {
       if (auto VD = dyn_cast<VarDecl>(D)) {
+        // If the type of the current object is marked as non_differentiable, skip generating its derivative
+        if (auto* typeDecl = VD->getType()->getAsCXXRecordDecl()) {
```
Wouldn't all the variables in a declaration statement be of the same type? If so, can this be moved outside the loop?

PhrygianGates replied:
I think so, this can be moved outside.

PhrygianGates replied:
Understood. Given that all variables within a single declaration statement can only have the same attribute, I have optimized the code to check only the first element in each declaration statement.

parth-07 replied:
Great!

parth-07 commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -1550,6 +1625,14 @@ namespace clad {
+        if (auto* typeDecl = VD->getType()->getAsCXXRecordDecl()) {
+          auto* Attr = typeDecl->getAttr<clang::AnnotateAttr>();
+          if (Attr && Attr->getAnnotation() == "non_differentiable") {
+            decls.push_back(VD);
```
Where is these decls values used?

PhrygianGates replied:
I think they are used later in line 1666.

parth-07 reviewed (changes requested):
Initial review.

PhrygianGates commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -213,6 +213,16 @@ namespace clad {
            "Doesn't support recursive diff. Use DiffPlan.");
     m_DerivativeInFlight = true;
 
+    if (const CXXMethodDecl* MD = dyn_cast<CXXMethodDecl>(m_Function)) {
```
We have reviewed all free functions in the `DerivativeBuilder::Derive` method. Upon reflection, I believe it may be more efficient to transfer the code from `ForwardModeVisitor::Derive` to `DerivativeBuilder::Derive`.

Checking for 'differentiable' in the `visit` function call methods seems impractical, given that an `CallExpr` won't frequently appear in the AST. We would only use the `VisitCallExpr` function when another function is invoked within this function. Consider the following example:

```cpp
double mem_fn_3(double i, double j)  { 
  SimpleFunctions obj(2, 3);
  return obj.mem_fn(i, j) + i*j; 
}
```

In this case, the `VisitCallExpr` function is used because `obj.mem_fn(i, j)` is a function call within `mem_fn_3`.

parth-07 commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -1319,6 +1333,58 @@ namespace clad {
       return StmtDiff(Clone(CE));
     }
+
+    // If the function is non_differentiable, return zero derivative.
```
Please refactor this whole code to be more modular. For example:
```cpp
if (HasNonDifferentiableAttribute(fn))
  return {Clone(CE), utils::getZero(...)};
```

parth-07 commented on test/ForwardMode/NonDifferentiable.C:
```diff
@@ -0,0 +1,155 @@
+  classname expr_1(2, 3);                                                      \
+  classname expr_2(3, 5);
+
+#define TEST_CLASS(classname, name, i, j)                                      \
```
Instead of creating new macros, can you please see if testing utility macros in `test/TestUtils.h` can be reused here?

PhrygianGates replied:
After reviewing the code in test/TestUtils.h, I guess the macros there cannot be reused in this context. Those macros are specifically designed for free functions, whereas our requirement is to use class methods.

parth-07 replied:
Okay, got it.

PhrygianGates replied:
> Okay, got it. Do you think if the TestUtils framework can be extended to work with class methods as well?

Apologies for overlooking this review earlier. If necessary, I can integrate the macro into the TestUtils framework.

parth-07 commented on test/ForwardMode/NonDifferentiable.C:
```diff
@@ -0,0 +1,155 @@
+  double x;
+  non_differentiable double y;
+  double mem_fn_1(double i, double j) { return (x + y) * i + i * j * j; }
+  non_differentiable double mem_fn_2(double i, double j) { return i * j; }
```
Can you please add a test where a non-differentiable function is being multiplied by another function and a variable?

PhrygianGates replied:
yeah, already done

parth-07 replied:
Can you please tell me which function in the test file adds this test case?

parth-07 commented on lib/Differentiator/DerivativeBuilder.cpp:
```diff
@@ -178,7 +178,35 @@ namespace clad {
+
+    // If the function is a method of a class, check if the class is
+    // non-differentiable.
+    if (const CXXMethodDecl* MD = dyn_cast<CXXMethodDecl>(FD)) {
```
What do you think about adding a general function, `hasNonDifferentiableAttribute`, that can be used on any kind of declaration?

PhrygianGates replied:
already done

parth-07 replied:
Okay, then why are we not using such a function here and other places?

parth-07 reviewed (changes requested):
Overall looks good.

vgvassilev reviewed (commented):
We should try to add tests for the cases codecov reports.

parth-07 commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -999,7 +999,21 @@ namespace clad {
-          ConstantFolder::synthesizeLiteral(m_Context.DoubleTy, m_Context, 0);
+          ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
+      // If non_differentiable attribute is present, return 0.
+      auto* Attr = ME->getMemberDecl()->getAttr<clang::AnnotateAttr>();
```
Can we replace this whole code with something much more modular and readable such as:

```cpp
if (hasNonDifferentiableAttr(ME))
  return {clonedME, zero};
```

parth-07 replied:
This has already been fixed it seems.

parth-07 commented on lib/Differentiator/DerivativeBuilder.cpp:
```diff
@@ -178,7 +179,33 @@ namespace clad {
+      const CXXRecordDecl* CD = MD->getParent();
+      if (clad::utils::hasNonDifferentiableAttribute(CD)) {
+        diag(DiagnosticsEngine::Error, MD->getLocation(),
+             "attempted differentiation of method '%0' in class '%1', which is "
```
Just to make sure, this would only produce an error if a user has explicitly requested derivative of a non-differentiable function, right? In particular, it should not produce an error if a non-differentiable function is being used in an another differentiable function.

For example:

```cpp
non_differentiable double nondiff_fn() { ... }
double diff_fn() { ...; nondiff_fn(); ...}

int main() {
  auto nondiff_fn_diff = clad::differentiate(nondiff_fn, ...); // This should produce an error.
  auto diff_fn_diff = clad::differentiate(diff_fn, ..); // This should not produce any error
}
```

PhrygianGates replied:
Yes! that complies with what I have implemented, and is included in the existing tests. 
In NonDifferentiable.C, we have the following code:

```C++
class SimpleFunctions1 {
public:
  SimpleFunctions1() noexcept : x(0), y(0) {}
  SimpleFunctions1(double p_x, double p_y) noexcept : x(p_x), y(p_y) {}
  double x;
  non_differentiable double y;
  double mem_fn_1(double i, double j) { return (x + y) * i + i * j * j; }
  non_differentiable double mem_fn_2(double i, double j) { return i * j; }
  double mem_fn_3(double i, double j) { return mem_fn_1(i, j) + i * j; }
  double mem_fn_4(double i, double j) { return mem_fn_2(i, j) + i * j; }
  double mem_fn_5(double i, double j) { return mem_fn_2(i, j) * mem_fn_1(i, j) * i; }
  SimpleFunctions1 operator+(const SimpleFunctions1& other) const {
    return SimpleFunctions1(x + other.x, y + other.y);
  }
};
```

The function `mem_fn_5` calls `mem_fn_2` and can be differentiated.

parth-07 replied:
Okay, thank you for the information.

parth-07 commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -999,7 +999,17 @@ namespace clad {
-          ConstantFolder::synthesizeLiteral(m_Context.DoubleTy, m_Context, 0);
+          ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
+      // If non_differentiable attribute is present, return 0.
+      if (clad::utils::hasNonDifferentiableAttribute(ME->getMemberDecl())) {
```
Can we check non differentiable attribute in both the member and the class in one function call, `clad::utils::hasNonDifferentiableAttribute(ME)`?

PhrygianGates replied:
already done

parth-07 commented on lib/Differentiator/ForwardModeVisitor.cpp:
```diff
@@ -1319,6 +1329,54 @@ namespace clad {
+    }
+
+    // If the base class is non_differentiable, return zero derivative.
+    if (const auto* ME = dyn_cast<CXXMemberCallExpr>(CE)) {
```
The if condition for `clad::utils::hasNonDifferentiableAttribute(FD)` and `clad::utils::hasNonDifferentiableAttribute(classType)` seems pretty much same. Can we reuse the code?

Also, if a class has a non_differentiable attribute, then should the `clad::utils::hasNonDifferentiableAttribute(FD)` function return true for the corresponding member functions of the class?

PhrygianGates replied:
Yes, I have already added another non-differentiable attribute for both CallExpr and MemberExpr. Currently, we will return true for the corresponding member functions of a NonDifferentiable class.

PhrygianGates commented:
Hi! I've added tests for error cases, but it appears that others have also modified the ForwardModeVisitor.cpp file. I am currently working on updating my code to fit the new code structure.

parth-07 commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,63 @@ namespace clad {
+
+    bool hasNonDifferentiableAttribute(const clang::Decl* D) {
+      auto* Attr = D->getAttr<clang::AnnotateAttr>();
+      return Attr && Attr->getAnnotation() == "non_differentiable";
```
This will give incorrect results if a declaration has multiple attributes and `non_differentiable` is not the 
first attribute of the declaration. A declaration can have multiple attributes. 
`Decl::getAttr<clang::AnnotateAttr` will only return the first attribute of a declaration.

Can you please fix this and add a test for the same?

PhrygianGates replied:
Thank you; I was unaware of it. I will promptly fix the issue.

parth-07 commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,63 @@ namespace clad {
+        const clang::ValueDecl* memberDecl = ME->getMemberDecl();
+        if(memberDecl) {
+          auto* Attr = memberDecl->getAttr<clang::AnnotateAttr>();
+          if(Attr && Attr->getAnnotation() == "non_differentiable") {
```
Same as before. It will give incorrect results if there are multiple attributes and `non_differentiable` is not the first
attribute of the declaration.

PhrygianGates replied:
already done

parth-07 commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,63 @@ namespace clad {
+        // Check attribute of the base class
+        if(auto classType = ME->getBase()->getType()->getAsCXXRecordDecl()) {
+          if(classType) {
+            auto* Attr = classType->getAttr<clang::AnnotateAttr>();
```
Same as before. It will give incorrect results if there are multiple attributes and `non_differentiable` is not the first
attribute of the class.

parth-07 commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,63 @@ namespace clad {
+        // Check if the function is non-differentiable
+        const clang::FunctionDecl* FD = CE->getDirectCallee();
+        if(FD) {
+          auto* Attr = FD->getAttr<clang::AnnotateAttr>();
```
Same as before. It will give incorrect results if there are multiple attributes and `non_differentiable` is not the first
attribute of the function.

parth-07 commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,63 @@ namespace clad {
+                                ->getType()
+                                ->getAsCXXRecordDecl()) {
+            if(classType) {
+              auto* Attr = classType->getAttr<clang::AnnotateAttr>();
```
Same as before. It will give incorrect results if there are multiple attributes and `non_differentiable` is not the first
attribute of the function.

parth-07 commented on lib/Differentiator/DerivativeBuilder.cpp:
```diff
@@ -180,7 +181,34 @@ namespace clad {
+    // check if the function is non-differentiable.
+    if (clad::utils::hasNonDifferentiableAttribute(FD)) {
+      if (request.VerboseDiags)
+        diag(DiagnosticsEngine::Error,
```
Why do we only print the error if verbose diagnostics is true?

PhrygianGates replied:
I'll remove the if verbose check.

parth-07 commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,30 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+  // If the DeclStmt is not empty, check the first declaration.
+  if (declsBegin != declsEnd && isa<VarDecl>(*declsBegin)) {
+    auto VD = dyn_cast<VarDecl>(*declsBegin);
+    auto* typeDecl = VD->getType()->getAsCXXRecordDecl();
```
Will this work correctly for the below declaration statement?

```cpp
NonDifferentiableClass *d1, d2, d3;
```

Here the first declaration is a pointer.

PhrygianGates replied:
It can't work correctly for the given declaration statement.

The type checking in this code is checking the CXXRecordDecl of the VarDecl type, and since pointers in C++ are fundamentally different from non-pointer types, typeDecl will be null for d1 because the getAsCXXRecordDecl() method will not return a record declaration for a pointer type. Therefore, this code would not correctly recognize d1 as a variable of a non-differentiable type, even though it's a pointer to such a type.

For d2 and d3, since the checking only happens on the first declaration (which is d1), these declarations won't be processed under the current implementation.

I'll fix this issue.

PhrygianGates commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -615,7 +615,10 @@ StmtDiff BaseForwardModeVisitor::VisitMemberExpr(const MemberExpr* ME) {
   } else {
     auto zero =
-        ConstantFolder::synthesizeLiteral(m_Context.DoubleTy, m_Context, 0);
+        ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
```
Regarding functionality, the use of IntTy 0 is acceptable because it behaves identically to DoubleTy 0 in both multiplication and addition contexts. Furthermore, upon reviewing the original code, I found that m_Context.IntTy is used consistently, whereas m_Context.DoubleTy isn't utilized.

parth-07 replied:
Okay, can you please open an issue to make the usage of zero type consistent throughout the codebase?

parth-07 reviewed (approved):
Looks good.

vgvassilev commented:
@PhrygianGates can you squash all commits into one and rebase?

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,33 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+  auto declsEnd = DS->decls().end();
+
+  // If the DeclStmt is not empty, check the first declaration.
+  if (declsBegin != declsEnd && isa<VarDecl>(*declsBegin)) {
```
Can we write this conditional as:
`if (VarDecl *VD = dyn_cast<VarDecl>(DS->decls().begin())`

That is would this work correctly if begin() == end()?

vgvassilev replied:
Fair enough.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,33 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+    auto* typeDecl = QT->getAsCXXRecordDecl();
+    if (typeDecl && clad::utils::hasNonDifferentiableAttribute(typeDecl)) {
+      for (auto D : DS->decls()) {
+        if (auto VD = dyn_cast<VarDecl>(D)) {
+          decls.push_back(VD);
+        } else {
+          diag(DiagnosticsEngine::Warning, D->getEndLoc(),
+               "Unsupported declaration");
+        }
```
```suggestion
        if (auto VD = dyn_cast<VarDecl>(D))
          decls.push_back(VD);
        else
          diag(DiagnosticsEngine::Warning, D->getEndLoc(),
               "Unsupported declaration");
```

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+  if (declsBegin != declsEnd && isa<VarDecl>(*declsBegin)) {
+    auto VD = dyn_cast<VarDecl>(*declsBegin);
+    // Check for non-differentiable types.
+    QualType QT = VD->getType();
```
@PhrygianGates, please do not address these comments. @vaithak is looking how to suppress them.

PhrygianGates replied:
ok

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+    if (QT->isPointerType())
+      QT = QT->getPointeeType();
+    auto* typeDecl = QT->getAsCXXRecordDecl();
+    if (typeDecl && clad::utils::hasNonDifferentiableAttribute(typeDecl)) {
```
@vaithak we should probably ignore such suggestions too.

vgvassilev replied:
We might tweak this check by adding `AllowPointerConditions=true`.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@
+    auto* typeDecl = QT->getAsCXXRecordDecl();
+    if (typeDecl && clad::utils::hasNonDifferentiableAttribute(typeDecl)) {
+      for (auto D : DS->decls()) {
+        if (auto VD = dyn_cast<VarDecl>(D))
```
We should ignore that too.

vgvassilev commented:
Can you rebase on top of master?

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -616,6 +616,9 @@ StmtDiff BaseForwardModeVisitor::VisitMemberExpr(const MemberExpr* ME) {
   } else {
     auto zero =
         ConstantFolder::synthesizeLiteral(m_Context.DoubleTy, m_Context, 0);
+    if (clad::utils::hasNonDifferentiableAttribute(ME)) {
+      return {clonedME, zero};
+    }
```
```suggestion
    if (clad::utils::hasNonDifferentiableAttribute(ME))
      return {clonedME, zero};
```

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -919,6 +922,27 @@ StmtDiff BaseForwardModeVisitor::VisitCallExpr(const CallExpr* CE) {
+  if (clad::utils::hasNonDifferentiableAttribute(CE)) {
+    // Calling the function without computing derivatives
+    llvm::SmallVector<Expr*, 4> ClonedArgs;
+    for (unsigned i = 0, e = CE->getNumArgs(); i < e; ++i) {
+      ClonedArgs.push_back(Clone(CE->getArg(i)));
+    }
```
```suggestion
    for (unsigned i = 0, e = CE->getNumArgs(); i < e; ++i)
      ClonedArgs.push_back(Clone(CE->getArg(i)));
```

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -919,6 +922,27 @@ StmtDiff BaseForwardModeVisitor::VisitCallExpr(const CallExpr* CE) {
+                                    noLoc, ClonedArgs, noLoc)
+                     .get();
+    // Creating a zero derivative
+    auto zero =
```
I think that suggestion is good and we should take it.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
   llvm::SmallVector<Decl*, 4> declsDiff;
+  // If the type is marked as non_differentiable, skip generating its derivative
+  // Get the iterator
+  auto declsBegin = DS->decls().begin();
```
Likewise.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+  // If the type is marked as non_differentiable, skip generating its derivative
+  // Get the iterator
+  auto declsBegin = DS->decls().begin();
+  auto declsEnd = DS->decls().end();
```
That too.

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+
+  // If the DeclStmt is not empty, check the first declaration.
+  if (declsBegin != declsEnd && isa<VarDecl>(*declsBegin)) {
+    auto VD = dyn_cast<VarDecl>(*declsBegin);
```
+1

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@ VarDeclDiff BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD) {
+    auto VD = dyn_cast<VarDecl>(*declsBegin);
+    // Check for non-differentiable types.
+    QualType QT = VD->getType();
+    if (QT->isPointerType())
+      QT = QT->getPointeeType();
```
Can you exclude `readability-braces-around-statements` from the clang-tidy checks in the `.clang-tidy` file?

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@
+      QT = QT->getPointeeType();
+    auto* typeDecl = QT->getAsCXXRecordDecl();
+    if (typeDecl && clad::utils::hasNonDifferentiableAttribute(typeDecl)) {
+      for (auto D : DS->decls()) {
```
+1

vgvassilev commented on lib/Differentiator/BaseForwardModeVisitor.cpp:
```diff
@@ -1322,6 +1346,32 @@
+    auto* typeDecl = QT->getAsCXXRecordDecl();
+    if (typeDecl && clad::utils::hasNonDifferentiableAttribute(typeDecl)) {
+      for (auto D : DS->decls()) {
+        if (auto VD = dyn_cast<VarDecl>(D))
```
+1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,68 @@ namespace clad {
     }
+
+    bool hasNonDifferentiableAttribute(const clang::Decl* D) {
+      for (auto* Attr : D->specific_attrs<clang::AnnotateAttr>()) {
+        if (Attr->getAnnotation() == "non_differentiable") {
+          return true;
+        }
+      }
```
Ping.

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+      // Check MemberExpr
+      if (const clang::MemberExpr* ME = clang::dyn_cast<clang::MemberExpr>(E)) {
+        // Check attribute of the member declaration
+        if (auto memberDecl = ME->getMemberDecl()) {
```
+1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+      if (const clang::MemberExpr* ME = clang::dyn_cast<clang::MemberExpr>(E)) {
+        // Check attribute of the member declaration
+        if (auto memberDecl = ME->getMemberDecl()) {
+          if (hasNonDifferentiableAttribute(memberDecl))
+            return true;
```
-1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+        }
+
+        // Check attribute of the base class
+        if (auto classType = ME->getBase()->getType()->getAsCXXRecordDecl()) {
```
+1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+
+        // Check attribute of the base class
+        if (auto classType = ME->getBase()->getType()->getAsCXXRecordDecl()) {
+          if (hasNonDifferentiableAttribute(classType))
+            return true;
```
-1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+                   clang::dyn_cast<clang::CallExpr>(E)) {
+        // Check if the function is non-differentiable
+        if (auto FD = CE->getDirectCallee()) {
+          if (hasNonDifferentiableAttribute(FD))
+            return true;
```
-1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+
+        // Check if the base class is non-differentiable
+        if (const auto* ME = dyn_cast<CXXMemberCallExpr>(CE)) {
+          if (auto classType = ME->getImplicitObjectArgument()
```
+1

vgvassilev commented on lib/Differentiator/CladUtils.cpp:
```diff
@@ -485,5 +485,53 @@ namespace clad {
+          if (auto classType = ME->getImplicitObjectArgument()
+                                   ->getType()
+                                   ->getAsCXXRecordDecl()) {
+            if (hasNonDifferentiableAttribute(classType))
+              return true;
```
-1

vgvassilev commented on lib/Differentiator/DerivativeBuilder.cpp:
```diff
@@ -180,7 +181,33 @@ namespace clad {
+    // check if the function is non-differentiable.
+    if (clad::utils::hasNonDifferentiableAttribute(FD)) {
+      diag(DiagnosticsEngine::Error,
+           request.CallContext ? request.CallContext->getBeginLoc() : noLoc,
```
-1

vgvassilev reviewed (approved):
LGTM! Please update the commit message to match the PR description and I can merge this.

vgvassilev commented:
> LGTM! Please update the commit message to match the PR description and I can merge this.

And please apply clang-format to the PR.

PhrygianGates commented:
> > LGTM! Please update the commit message to match the PR description and I can merge this.
> 
> And please apply clang-format to the PR.

ok

vgvassilev commented:
This looks good, however, when we are not rebased on top of the master the git-clang-format picks up other issues and reports them. @vaithak we should either automatically rebase the PR which is hard, or pick up the changes done by the author only. I've done that [here](https://github.com/root-project/root/blob/063e1b1a8fd2d32479e98969b3bac258b6909c7d/.ci/format_script.sh#L5-L7) long time ago and I think the approach still works. Can you adapt this to fit out CI?

vaithak commented:
> This looks good, however, when we are not rebased on top of the master the git-clang-format picks up other issues and reports them. @vaithak we should either automatically rebase the PR which is hard, or pick up the changes done by the author only. I've done that [here](https://github.com/root-project/root/blob/063e1b1a8fd2d32479e98969b3bac258b6909c7d/.ci/format_script.sh#L5-L7) long time ago and I think the approach still works. Can you adapt this to fit out CI?

I experienced this too in the last PR, will look into this :+1:
